main/init/sbt/models/src/mill/main/sbt/Models.scala

```diff
@@ -0,0 +1,171 @@
+package mill.main.sbt
+
+import mill.main.sbt.BuildPublicationInfo.License
+import mill.main.sbt.Models.URL
+import sbt.Keys
+import upickle.default.{macroRW, ReadWriter => RW}
+
+object Models {
+  type URL = String
+}
+
+case class BuildExport(
+    /** @see [[sbt.AutoPlugin.buildSettings]] and [[sbt.AutoPlugin.globalSettings]] */
+    defaultBuildInfo: BuildInfo,
+    projects: Seq[Project]
+)
+object BuildExport {
+  implicit val rw: RW[BuildExport] = macroRW
+}
+
+case class BuildInfo(
+    buildPublicationInfo: BuildPublicationInfo,
+    /** @see [[Keys.javacOptions]] */
+    javacOptions: Option[Seq[String]],
+    /** @see [[Keys.scalaVersion]] */
+    scalaVersion: Option[String],
+    /** @see [[Keys.scalacOptions]] */
+    scalacOptions: Option[Seq[String]],
+    /** @see [[Keys.resolvers]] */
+    resolvers: Option[Seq[Resolver]]
+)
+object BuildInfo {
+  implicit val rw: RW[BuildInfo] = macroRW
+}
+
+/**
+ * Members ordered by their order in [[Keys]].
+ */
+case class BuildPublicationInfo(
+    /** @see [[Keys.description]] */
+    description: Option[String],
+    /**
+     * corresponds to `url` in POM
+     *
+     * @see [[Keys.homepage]]
+     */
+    homepage: Option[Option[String]],
+    /** @see [[Keys.licenses]] */
+    licenses: Option[Seq[License]],
+    /**
+     * corresponds to `groupId` in POM and Mill's `PomSettings.organization`
+     *
+     * @see [[Keys.organization]]
+     */
+    organization: Option[String],
+    // not needed
+    /*
+    /**
+     * corresponds to Maven's `organization` in POM
+     *
+     * @see [[Keys.organizationName]]
+     */
+    organizationName: Option[String],
+     */
+    // not needed
+    /*
+    /**
+     * corresponds to `organizationUrl` in POM
+     *
+     * @see [[Keys.organizationHomepage]]
+     */
+    organizationHomepage: Option[Option[String]],
+     */
+    /** @see [[Keys.developers]] */
+    developers: Option[Seq[Developer]],
+    /** @see [[Keys.scmInfo]] */
+    scmInfo: Option[Option[ScmInfo]],
+    /** @see [[Keys.version]] */
+    version: Option[String]
+)
+object BuildPublicationInfo {
+
+  /** @see [[sbt.librarymanagement.License]] */
+  type License = (String, URL)
+
+  implicit val rw: RW[BuildPublicationInfo] = macroRW
+}
+
+/**
+ * @see [[sbt.librarymanagement.ScmInfo]]
+ */
+case class ScmInfo(
+    browseUrl: URL,
+    connection: String,
+    devConnection: Option[String]
+)
+object ScmInfo {
+  implicit val rw: RW[ScmInfo] = macroRW
+}
+
+/**
+ * @see [[sbt.librarymanagement.Developer]]
+ */
+case class Developer(id: String, name: String, email: String, url: URL)
```

Comment:
This seems to duplicate a bunch of stuff in `ir.scala`; can we re-use some of that? Or does the data model diverge enough it can't work?

Reply:
I think we call reuse some but others diverge. For example, `organization` refers to the Maven group in both sbt and Mill, but it refers to the name of the organization in Maven and Gradle, equivalent to sbt's `organizationName`.

Another difficulty here is that the IR models depend on some other modules and libraries such as `build.runner` and OS-Lib, and don't have uPickle configured.

Also to reduce duplication I think it's also needed to remove and unify some Java models in the Gradle module, as I commented last time reviewing the Gradle PR.

So in overall, I think this may involve some work can come in a cleanup and refactor PR after this.